Mainnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs

```diff
@@ -0,0 +1,639 @@
+    private ulong GetTokenByIndex(ulong index) => this.PersistentState.GetUInt64(GetTokenByIndexKey(index));
+
+    private void SetTokenByIndex(ulong index, ulong token) => this.PersistentState.SetUInt64(GetTokenByIndexKey(index), token);
+
+    private void ClearTokenByIndex(ulong index) => this.PersistentState.Clear(GetTokenByIndexKey(index));
+
+    private string GetIndexByTokenKey(ulong token) => $"IndexByToken:{token}";
+
+    private ulong GetIndexByToken(ulong token) => this.PersistentState.GetUInt64(GetIndexByTokenKey(token));
+
+    private void SetIndexByToken(ulong token, ulong index) => this.PersistentState.SetUInt64(GetIndexByTokenKey(token), index);
+
+    private void ClearIndexByToken(ulong token) => this.PersistentState.Clear(GetIndexByTokenKey(token));
+
+    private string GetTokenOfOwnerByIndexKey(Address address, ulong index) => $"TokenOfOwnerByIndex:{address}:{index}";
+
+    private ulong GetTokenOfOwnerByIndex(Address address, ulong index) => this.PersistentState.GetUInt64(GetTokenOfOwnerByIndexKey(address, index));
+
+    private void SetTokenOfOwnerByIndex(Address owner, ulong index, ulong tokenId) => this.PersistentState.SetUInt64(GetTokenOfOwnerByIndexKey(owner, index), tokenId);
+
+    private void ClearTokenOfOwnerByIndex(Address owner, ulong index) => this.PersistentState.Clear(GetTokenOfOwnerByIndexKey(owner, index));
+
+    private string IndexOfOwnerByTokenKey(Address owner, ulong tokenId) => $"IndexOfOwnerByToken:{owner}:{tokenId}";
+    private ulong GetIndexOfOwnerByToken(Address owner, ulong tokenId) => this.PersistentState.GetUInt64(IndexOfOwnerByTokenKey(owner, tokenId));
+    private void SetIndexOfOwnerByToken(Address owner, ulong tokenId, ulong index) => this.PersistentState.SetUInt64(IndexOfOwnerByTokenKey(owner, tokenId), index);
+    private void ClearIndexOfOwnerByToken(Address owner, ulong tokenId) => this.PersistentState.Clear(IndexOfOwnerByTokenKey(owner, tokenId));
+    public ulong TotalSupply
+    {
+        get => this.PersistentState.GetUInt64(nameof(TotalSupply));
+        private set => this.PersistentState.SetUInt64(nameof(TotalSupply), value);
+    }
+
+    /// <summary>
+    /// Constructor. Initializes the supported interfaces.
+    /// </summary>
+    /// <param name="state">The smart contract state.</param>
+    public NonFungibleToken(ISmartContractState state, string name, string symbol) : base(state)
+    {
+        // todo: discuss callback handling and supported interface numbering with community.
+        this.SetSupportedInterfaces((uint)0x00000001, true); // (ERC165) - ISupportsInterface
+        this.SetSupportedInterfaces((uint)0x00000002, true); // (ERC721) - INonFungibleToken,
+        this.SetSupportedInterfaces((uint)0x00000003, false); // (ERC721) - INonFungibleTokenReceiver
+        this.SetSupportedInterfaces((uint)0x00000004, true); // (ERC721) - INonFungibleTokenMetadata
+        this.SetSupportedInterfaces((uint)0x00000005, true); // (ERC721) - IERC721Enumerable
```

Comment:
Is there a practical use case for these interfaces? Normally in ETH you'd cast or tell the contract what interface another contract should be using. 

This sounds like ContractA interacting with this contract (Contract B) would have to make a call to check what interface is supported _before_ interacting with the contract. 

This is where whitelisted interfaces come into play.

ContractA should be able to say Assert(ContractB is ISomeInterface). I remember @Neurosploit I think it was that initially started this contract. Maybe he'll have some feedback.

Reply:
@mrtpain @YakupIpek  See the PR here https://github.com/stratisproject/StratisSmartContractsSamples/pull/7
That's as far as it has gotten and has left my memory completely so I can't be of much help.

Reply:
Unfortunately `Assert(ContractB is ISomeInterface)` this type of casting is not yet possible. I guess early version of solidity didn't support it as well so that is why the original code include alternative way to call SupportInterface method.

Reply:
should INonFungibleTokenReceiver always be false? Is there a reason to include it if so?

Reply:
I will keep it as explicit documentation for now.